feat: enable custom password and generate access token during initialization for ff-admin

AllanOricil opened the pull request:
## Goal

Enable users to programmatically manage their FF instance resources using the Rest API without requiring them to manually complete the setup, or create an admin user through the UI.

## Description

With these changes:
- `ff-admin` is created during start up.
- `ff-admin` password can be set using `FF_ADMIN_PASSWORD` env variable instead of being random.
- `ff-admin` personal access token can be created at start up using a new flow forge config called `create_admin_access_token`.
- when `create_admin = true` the setup is done automatically

I decided to move the `ff-admin` user creation from `./forge/routes/setup/index.js` to `./forge/forge.js` because the `ff-admin` creation no longer happens after accessing the `/setup` route, but when a forge server is started for the first time with no users.

## BEFORE
<img width="894" alt="Screenshot 2023-10-25 at 16 31 16" src="https://github.com/FlowFuse/flowfuse/assets/55927613/3968aa92-6c75-4c54-9c75-7c5914b6b0cc">


## AFTER

![image](https://github.com/FlowFuse/flowfuse/assets/55927613/6b402548-9c1f-4c2e-b658-031efbb20b5e)




## Related Issue(s)

N/A 
we discussed it on slack

## Checklist

<!-- https://flowfuse.com/handbook/development/#defining-done -->

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass <!-- If not adding test coverage, please clarify why not? -->
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

AllanOricil commented:
@hardillb I've fixed all lint issues: 

<img width="1033" alt="image" src="https://github.com/FlowFuse/flowfuse/assets/55927613/2eed6b5b-f700-4155-b11b-f946e17e7294">

knolleary commented:
I'm not sure we want this to be the default behaviour for a new install - which your changes to `etc/flowforge.yml` imply they would be.

You second screenshot also shows some of the fields it logs are mixed up - refresh_token and expires at. 

We also need to think about your prosed moving of this logic to the main start up rather than the setup route and what consequences that has. I'm not going to rush this into today's release, but once we resolve things, it can get into the next maintenance release.

AllanOricil commented:
@knolleary just pushed changes to fix what you mentioned

AllanOricil commented:
I've rebased it

AllanOricil commented:
@knolleary @hardillb is there a problem with these changes?

knolleary commented:
Sorry @AllanOricil, Ben and I have both been travelling this week. We'll get this PR reviewed properly asap.

hardillb commented:
Just looking this over, if we have both a admin created and a admin API token created, do we want to run `await app.settings.set('setup:initialised', true)` to disable the setup wizard?

This would only if both those options are enabled as we do not expect the user to access the web interface to complete the setup?

Otherwise it looks OK, will need to check it still applies cleanly

AllanOricil commented:
@hardillb 

One of the steps of the setup allows people to add a license. Is there another way to input it later?

<img width="852" alt="image" src="https://github.com/FlowFuse/flowfuse/assets/55927613/61e87d22-2a91-406d-a884-56ea4f4dcc4f">

hardillb commented:
Yes an admin can enter a license under the admin settings page

AllanOricil commented:
I'm going to add a condition to avoid running that piece of code you shared @hardillb Is that OK?

hardillb commented:
@AllanOricil I think we need to have an internal discussion about this first. It's a quick change we can add on if needed.

hardillb commented:
OK, having discussed this, calling `app.settings.set('setup:initialised', true)` is not going to work.

What will be better is to enable the token based auth for the last step of the setup wizard, which generates the default stacks and generates some other internal values that are needed.

This would mean that the first API call after the startup would be to complete setup properly which will in turn disable the setup wizard.

AllanOricil commented:
could you move the user creation call to the right place or show the line where it has to be moved to?

AllanOricil commented:
I think I know where to do it. I will finish it this weekend.

hardillb commented:
> This would mean that the first API call after the startup would be to complete setup properly which will in turn disable the setup wizard.

By this we mean after start up for the first time (when the user and API token are generated) the first API call should be to `/setup/finish` which creates the default instance type, stack and template and team type 

https://github.com/FlowFuse/flowfuse/blob/2004f93287e69712235317225d00cbcd8bd88de1/forge/routes/setup/index.js#L67

Which looking at the code doesn't need any authentication so there shouldn't need to be a code change

knolleary commented on forge/forge.js:
```diff
@@ -19,6 +21,38 @@ const settings = require('./settings')
+}
+
+async function createAdminAccessToken (server, userId) {
+    const { token, expiresAt, refreshToken } = await server.db.controllers.AccessToken.createTokenForUser(userId, null, ['all'], true)
+    server.log.info(`[SETUP] token: ${token}`)
+    server.log.info(`[SETUP] refresh token: ${refreshToken}`)
+    server.log.info(`[SETUP] expires at: ${expiresAt}`)
```
```suggestion
    const { token, expiresAt, refreshToken } = await server.db.controllers.AccessToken.createPersonalAccessToken(userId, '', null, 'Admin Access Token')
    server.log.info(`[SETUP] token: ${token}`)
```

AllanOricil replied:
done

AllanOricil commented:
> > This would mean that the first API call after the startup would be to complete setup properly which will in turn disable the setup wizard.
> 
> By this we mean after start up for the first time (when the user and API token are generated) the first API call should be to `/setup/finish` which creates the default instance type, stack and template and team type
> 
> https://github.com/FlowFuse/flowfuse/blob/2004f93287e69712235317225d00cbcd8bd88de1/forge/routes/setup/index.js#L67
> 
> Which looking at the code doesn't need any authentication so there shouldn't need to be a code change

this is done as well @hardillb

AllanOricil commented on forge/setup.js:
I had to create this new file to avoid circular dependency. Both forge.js and routes/setup/setup.js need this logic.

AllanOricil commented:
Looking forward to becoming a contributor. The first Brazilian to contribute? Can I add it to me resume? If u give me a medal I can maybe solve some open issues for you haha I'm kidding. I do it mostly for curiosity and to learn.

hardillb commented on test/system/002-setup-admin_spec.js:
```diff
@@ -8,47 +11,32 @@ describe('Default Admin User', function () {
     // inbox - a local transport we can use to capture email without an SMTP server
     const inbox = new LocalTransport()
 
-    before(async function () {
```
Why have you removed the `before` function? this is needed to reset the test environment.

AllanOricil replied:
It is because I'm creating flowfuse inside each test method with different configs. I could not find another way.

hardillb replied:
If you need to totally reset the environment between each test then the `before()` and `after()` functions should be renamed to `beforeEach()` and `afterEach()`

hardillb replied:
OK, this can probably stay as it is, but the await in the `after()` will be needed

hardillb commented:
This is failing tests, most likely because of the removed `before()` function.

Also did you rebase on main before the last update?

Thanks.

AllanOricil commented:
> This is failing tests, most likely because of the removed `before()` function.
> 
> Also did you rebase on main before the last update?
> 
> Thanks.

I merged main to the branch. Let me rebase it, instead.

hardillb commented on test/system/002-setup-admin_spec.js:
```diff
@@ -8,47 +11,32 @@ describe('Default Admin User', function () {
+    }
 
     after(function () {
         return forge.close()
```
```suggestion
        return await forge.close()
```

This is probably the problem, you need to wait for the close to finish before the next test

hardillb replied:
Ahh, this will also need a change to line 34 as well (which I can't do as a suggestion)

It will need to be:

```
    after( async function () {
        return await forge.close()
}
```

AllanOricil commented:
@hardillb try again. I changed after to afterEach, since every test has its own forge instance. I have also added the missing "async" keyword. 

<img width="1552" alt="image" src="https://github.com/FlowFuse/flowfuse/assets/55927613/f9056bc7-7991-4248-9d78-577b9cc6be7c">
<img width="1546" alt="image" src="https://github.com/FlowFuse/flowfuse/assets/55927613/a06885b1-5e19-4b7f-854d-980c3d7b2562">

AllanOricil commented:
I'm excited and emotive right now  🥺

hardillb commented on forge/forge.js:
```diff
@@ -15,9 +17,40 @@ const license = require('./licensing')
+    server.log.info('[SETUP] username: ff-admin')
+    server.log.info(`[SETUP] password: ${password}`)
+
+    if (server.config.create_admin_access_token) await createAdminAccessToken(server, userId)
```
```suggestion
    if (server.config.create_admin_access_token) {
        await createAdminAccessToken(server, userId)
    }
```

Just a style thing

hardillb commented on forge/forge.js:
```diff
@@ -15,9 +17,40 @@ const license = require('./licensing')
+}
+
+async function createAdminUser (server) {
+    if (await server.db.models.User.count() !== 0) return
```
```suggestion
    if (await server.db.models.User.count() !== 0) {
       return
    }
```

Just a style thing

hardillb reviewed (changes requested):
Looks good, one minor style thing, I prefer `if` blocks to to have curly braces as it makes them more obvious and easier to read.

And we could also do with another line adding to the `docs/install/configuration.md` file to describe the `create_admin_access_token` option under `create_admin` entry

AllanOricil commented:
Anything else master @hardillb ?

AllanOricil commented:
is it possible to exclude docs from codecov analysis? It failed because coverage decreased 0.04% after I added a new line in one of the docs

AllanOricil commented:
It is https://docs.codecov.com/docs/ignoring-paths

You can do it on the project config in codecov's UI, or in a codecov.yml file in the project

hardillb commented:
The docs are not covered by the codecov stuff, don't worry about it.

We will get this wrapped up this week, just been busy today

hardillb commented:
@knolleary I'm happy with this if you want to give it a quick look.

AllanOricil commented:
@hardillb @knolleary I think we need a definition of a super admin because ff-admin can be deleted by any other admin. If an admin deleted ff-admin, then we lose control over the instance.

<img width="1623" alt="image" src="https://github.com/FlowFuse/flowfuse/assets/55927613/8ccbb509-b94d-47c2-b29c-cd27c2c1d85b">

hardillb commented:
This is not related to this PR, but it is not possible to delete the last admin, so it is not possible to have no admin and  you should only promote trusted users to be an admin

AllanOricil commented:
🤨
